certs/certs.go

```diff
@@ -0,0 +1,293 @@
+package certs
+
+import (
+	"bytes"
+	"cmp"
+	"slices"
+	"sort"
+
+	"github.com/filecoin-project/go-bitfield"
+	"github.com/filecoin-project/go-f3/gpbft"
+	"golang.org/x/xerrors"
+)
+
+// PowerTableDelta represents a single power table change between GPBFT instances.
+type PowerTableDelta struct {
+	// Participant with changed power
+	ParticipantID gpbft.ActorID
+	// Change in power from base (signed).
+	PowerDelta *gpbft.StoragePower
+	// New signing key if relevant (else empty)
+	SigningKey gpbft.PubKey
+}
+
+func (d *PowerTableDelta) IsZero() bool {
+	return d.PowerDelta.Sign() == 0 && len(d.SigningKey) == 0
+}
+
+// FinalityCertificate represents a single finalized GPBFT instance.
+type FinalityCertificate struct {
+	// The GPBFT instance to which this finality certificate corresponds.
+	GPBFTInstance uint64
+	// The ECChain finalized during this instance, starting with the last tipset finalized in
+	// the previous instance.
+	ECChain gpbft.ECChain
+	// Additional data signed by the participants in this instance. Currently used to certify
+	// the power table used in the next instance.
+	SupplementalData gpbft.SupplementalData
+	// Indexes in the base power table of the certifiers (bitset)
+	Signers bitfield.BitField
+	// Aggregated signature of the certifiers
+	Signature []byte
+	// Changes between the power table used to validate this finality certificate and the power
+	// used to validate the next finality certificate. Sorted by ParticipantID, ascending.
+	PowerTableDelta []PowerTableDelta
+}
+
+// NewFinalityCertificate constructs a new finality certificate from the given power delta (from
+// `MakePowerTableDiff`) and justification (from GPBFT).
+//
+// Note, however, that this function does not attempt to validate the resulting finality
+// certificate (beyond verifying that it is a justification for the correct round). You can do so by
+// immediately calling `ValidateFinalityCertificates` on the result.
+func NewFinalityCertificate(powerDelta []PowerTableDelta, justification *gpbft.Justification) (*FinalityCertificate, error) {
+	if justification.Vote.Step != gpbft.DECIDE_PHASE {
+		return nil, xerrors.Errorf("can only create a finality certificate from a decide vote, got phase %s", justification.Vote.Step)
+	}
+
+	if justification.Vote.Round != 0 {
+		return nil, xerrors.Errorf("expected decide round to be 0, got round %d", justification.Vote.Round)
+	}
+
+	if justification.Vote.Value.IsZero() {
+		return nil, xerrors.Errorf("got a decision for bottom for instance %d", justification.Vote.Instance)
+	}
```

Comment:
More preconditions you might check:
- Vote value is not empty, and does Validate()
- Signers is not empty?
- Supplemental Merkel hash is the right length?

Edit: I see some of these are captured below, but perhaps some of these easier, pure-syntax ones would be better earlier?

Reply:
I can but, IMO, the best solution is just to call `ValidateFinalityCertificates` on the result. It's unclear where to draw the line:

- Validate the signature? Requires the verifier and network name.
- Check the supplemental data? Requires signature validation.
- Check that we have enough power? Requires the power table.
- Validate the bitfield?
- Validate the power delta? Requires the original power table.

So instead I opted to validate information that _won't_ end up in the finality certificate (i.e., make sure the justification is actually for a decide). Well, that and the fact that we're voting on _something_.

Options are:

1. Separate out a "syntactic validation" function from `ValidateFinalityCertificate` in both places.
2. Call `ValidateFinalityCertificates` inside `NewFinalityCertificate` (need to pass a power table, network name, etc.).
3. Tell the user to call `ValidateFinalityCertificates`.

IMO, we should be calling `ValidateFinalityCertificates` regardless which reduces the usefulness of 1. However, I hesitate to go with 2 because it makes it impossible to construct a finality certificate from a justification without a verifier, etc.